tests/functional/test_paramfile.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import json
+import logging
+
+from mock import patch, Mock, mock_open
+
+from awscli.testutils import FileCreator, BaseCLIWireResponseTest
+from awscli.clidriver import create_clidriver
+from botocore.vendored import requests
+
+logger = logging.getLogger(__name__)
+
+
+class FakeResponse(object):
+    def __init__(self, *args, **kwargs):
+        self.status_code = 200
+        self.text = 'function-name'
+
+
+class BaseTestCLIFollowParamURL(BaseCLIWireResponseTest):
+    def setUp(self):
+        super(BaseTestCLIFollowParamURL, self).setUp()
+        self.wire_response = json.dumps({}).encode('utf-8')
+        self.patch_send(content=self.wire_response)
+        self.files = FileCreator()
+        self.command_prefix = ['lambda', 'get-function', '--function-name']
+        self.start_patches()
+
+    def tearDown(self):
+        super(BaseTestCLIFollowParamURL, self).tearDown()
+        self.files.remove_all()
+        self.stop_patches()
+
+    def start_patches(self):
+        self.mock_open = mock_open(read_data='function-name')
+        self.open_patch = patch(
+            'awscli.paramfile.compat_open', self.mock_open
+        )
+        self.open_patch.start()
+
+        self.mock_get = Mock(spec=requests.get, side_effect=FakeResponse)
+        self.requests_patch = patch(
+            'awscli.paramfile.requests.get', self.mock_get
+        )
+        self.requests_patch.start()
+
+    def stop_patches(self):
+        self.requests_patch.stop()
+        self.open_patch.stop()
+
+    def run_cmd_with_arg(self, argument):
+        command = self.command_prefix + [argument]
+        self.run_cmd(command)
+
+
+class TestCLIFollowParamURLDefault(BaseTestCLIFollowParamURL):
```

Comment:
Hmmm. I think for these functional tests it would be better if we minimized the amount of assertions on mock we are making and be more explicit in each test case in what the inputs and outputs are. So specifically, I think we should try to copy the same style we do for the service-specific functional tests:
```python
class TestCLIFollowParamURLDefault(BaseTestCLIFollowParamURL):
    def test_with_http_prefix(self):
        command = "lambda get-function --function-name http://myremote.url"
        self.parsed_response = {}
        with mock.patch('awscli.paramfile.requests.get', spec=True) as mock_get:
            mock_get = FakeResponse(content='my-function-name')
            self.assert_params_for_cmd(command, {'FunctionName': 'my-function-name'}
            mock_get.assert_called_with('http://myremote.url')
```
Sure it is a little longer than what you currently have, but I find it a lot easier to read and understand. It also asserts that paramfile value were properly pulled into the client parameters. Would like to know your thoughts.